buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java

```diff
@@ -0,0 +1,173 @@
+        final Map<String, File> seen = new TreeMap<>();
+
+        licensesDirs.forEach(file -> {
+            try (Stream<Path> pathStream = Files.walk(file.toPath())) {
+                pathStream
+//                    .filter(path -> path.endsWith("-NOTICE.txt"))
+                    .filter(path -> path.toString().endsWith("-NOTICE.txt"))
+                    .map(Path::toFile)
+                    .forEach(licenseFile -> {
+                        // Here we remove the "-NOTICE.txt" so the name can be used later for the -LICENSE.txt file
+                        final String name =
+                            licenseFile.getName().substring(0, licenseFile.getName().length() - "-NOTICE.txt".length());
+
+                        if (seen.containsKey(name)) {
+                            File prevLicenseFile = seen.get(name);
+                            if (getText(prevLicenseFile).equals(getText(licenseFile)) == false) {
+                                throw new RuntimeException("Two different notices exist for dependency '" +
+                                    name + "': " + prevLicenseFile + " and " + licenseFile);
+                            }
+                        } else {
+                            seen.put(name, licenseFile);
+                        }
+                    });
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        for (Map.Entry<String, File> entry : seen.entrySet()) {
+            final String name = entry.getKey();
+            final File file = entry.getValue();
+            final File licenseFile = new File(file.getParentFile(), name + "-LICENSE.txt");
+
+            appendFileToOutput(file, name, "NOTICE", output);
+            appendFileToOutput(licenseFile, name, "LICENSE", output);
+        }
+
+        Files.write(outputFile.toPath(), output.toString().getBytes());
+    }
+
+    private static void appendFileToOutput(File file, final String name, final String type, StringBuilder output) {
+        String text = getText(file);
+
+        if (text.trim().isEmpty() == false) {
+            output.append("================================================================================\n");
+            output.append(name + " " + type + "\n");
+            output.append("================================================================================\n");
+            output.append(text);
+            output.append("\n\n");
+        }
+    }
+
+
+    private static String getText(File file) {
+        final StringBuilder builder = new StringBuilder();
+
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            char[] buffer = new char[8192];
+            int read;
+
+            while ((read = reader.read(buffer)) != -1) {
+                builder.append(buffer, 0, read);
+            }
+
+        } catch (IOException e) {
```

Comment:
It's generally not desirable to catch `IOException` low level, these should be bubbled up.